File: controller/internal/journal/journal.go
// Package journal provides the controller's append-only recovery journal.
//
// Journal records contain transition/runtime metadata only. Credential bytes,
// auth.json contents, authorization headers, and recovery prompt text are not
// part of Event and therefore cannot be serialized accidentally through the
// normal API.
package journal

import (
	"bufio"
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

var (
	ErrInvalidEvent   = errors.New("invalid journal event")
	ErrCorruptJournal = errors.New("corrupt journal")
	ErrSensitiveEvent = errors.New("journal event contains sensitive material")
)

// EventType is intentionally open to forward-compatible runtime events while
// the constants below document the MVP recovery lifecycle.
type EventType string

const (
	TransitionCreated    EventType = "TransitionCreated"
	TransitionPrepared   EventType = "TransitionPrepared"
	AuthDeployed         EventType = "AuthDeployed"
	CommitSent           EventType = "CommitSent"
	IdentityConfirmed    EventType = "IdentityConfirmed"
	TransitionCommitted  EventType = "TransitionCommitted"
	TransitionUncertain  EventType = "TransitionUncertain"
	TransitionReconciled EventType = "TransitionReconciled"
	RuntimeConnected     EventType = "RuntimeConnected"
	RuntimeDisconnected  EventType = "RuntimeDisconnected"
	// Recovery records describe the controller's durable authorization of a
	// runtime-owned UsageLimitExceeded continuation.  They contain only
	// correlation metadata; the prompt and credentials remain in Codext.
	RecoveryParked           EventType = "RecoveryParked"
	RecoveryWaiting          EventType = "RecoveryWaiting"
	RecoveryBound            EventType = "RecoveryBound"
	RecoveryReleaseRequested EventType = "RecoveryReleaseRequested"
	RecoveryReleased         EventType = "RecoveryReleased"
	RecoveryStarted          EventType = "RecoveryStarted"
	RecoveryCompleted        EventType = "RecoveryCompleted"
	RecoveryUncertain        EventType = "RecoveryUncertain"

	// Event-prefixed aliases make call sites self-documenting when journal and
	// runtime event constants are imported together.
	EventTransitionCreated        = TransitionCreated
	EventTransitionPrepared       = TransitionPrepared
	EventAuthDeployed             = AuthDeployed
	EventCommitSent               = CommitSent
	EventIdentityConfirmed        = IdentityConfirmed
	EventTransitionCommitted      = TransitionCommitted
	EventTransitionUncertain      = TransitionUncertain
	EventTransitionReconciled     = TransitionReconciled
	EventRuntimeConnected         = RuntimeConnected
	EventRuntimeDisconnected      = RuntimeDisconnected
	EventRecoveryParked           = RecoveryParked
	EventRecoveryWaiting          = RecoveryWaiting
	EventRecoveryBound            = RecoveryBound
	EventRecoveryReleaseRequested = RecoveryReleaseRequested
	EventRecoveryReleased         = RecoveryReleased
	EventRecoveryStarted          = RecoveryStarted
	EventRecoveryCompleted        = RecoveryCompleted
	EventRecoveryUncertain        = RecoveryUncertain
)

// Event is the complete journal wire record. The fields are deliberately
// scalar and non-secret; use Reason for a short operator-safe explanation, not
// raw provider responses or token material.
type Event struct {
	At           time.Time `json:"at"`
	Type         EventType `json:"type"`
	TransitionID string    `json:"transition_id,omitempty"`
	RecoveryID   string    `json:"recovery_id,omitempty"`
	ThreadID     string    `json:"thread_id,omitempty"`
	TurnID       string    `json:"turn_id,omitempty"`
	// Generation zero is a valid initial runtime generation. Keep the field on
	// the JSON record even when it is zero so every transition record carries
	// the correlation field required by the protocol contract.
	AuthGeneration     uint64 `json:"auth_generation"`
	ExpectedGeneration uint64 `json:"expected_generation,omitempty"`
	RuntimeID          string `json:"runtime_id,omitempty"`
	AccountID          string `json:"account_id,omitempty"`
	FromAccountID      string `json:"from_account_id,omitempty"`
	TargetAccountID    string `json:"target_account_id,omitempty"`
	Outcome            string `json:"outcome,omitempty"`
	Reason             string `json:"reason,omitempty"`
}

// Journal is the persistence seam consumed by transition orchestration.
type Journal interface {
	Append(event Event) error
	ReadAll() ([]Event, error)
}

// FileJournal is a process-safe JSONL journal. O_APPEND plus a process mutex
// keeps each record whole for concurrent callers in this process. Every append
// is synced before returning so a successful call is durable to the OS.
type FileJournal struct {
	path string
	mu   sync.Mutex
	now  func() time.Time
}

// NewFileJournal constructs a journal without creating its file.
func NewFileJournal(path string) *FileJournal {
	return &FileJournal{path: path, now: time.Now}
}

// NewJournal is the concise constructor used by controller wiring.
func NewJournal(path string) *FileJournal {
	return NewFileJournal(path)
}

// Path returns the journal path without exposing its records.
func (j *FileJournal) Path() string {
	if j == nil {
		return ""
	}
	return j.path
}

// SetClock injects a clock for deterministic tests. Event.At values supplied
// by callers remain authoritative; the clock fills only zero timestamps.
func (j *FileJournal) SetClock(now func() time.Time) {
	if j == nil {
		return
	}
	j.mu.Lock()
	defer j.mu.Unlock()
	if now == nil {
		j.now = time.Now
		return
	}
	j.now = now
}

// Append validates and durably appends one event. It never rewrites or prunes
// existing records.
func (j *FileJournal) Append(event Event) error {
	if j == nil {
		return errors.New("nil journal")
	}
	j.mu.Lock()
	defer j.mu.Unlock()
	if event.At.IsZero() {
		now := j.now
		if now == nil {
			now = time.Now
		}
		event.At = now()
	}
	event.At = event.At.UTC()
	if err := ValidateEvent(event); err != nil {
		return err
	}
	if j.path == "" {
		return errors.New("journal path is empty")
	}
	if err := rejectJournalSymlink(j.path); err != nil {
		return err
	}
	dir := filepath.Dir(j.path)
	if err := os.MkdirAll(dir, 0o700); err != nil {
		return err
	}
	if err := os.Chmod(dir, 0o700); err != nil && !isPermissionMetadataError(err) {
		return err
	}
	raw, err := json.Marshal(event)
	if err != nil {
		return err
	}
	raw = append(raw, '\n')
	file, err := os.OpenFile(j.path, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o600)
	if err != nil {
		return err
	}
	defer file.Close()
	if err := file.Chmod(0o600); err != nil && !isPermissionMetadataError(err) {
		return err
	}
	if _, err := file.Write(raw); err != nil {
		return err
	}
	if err := file.Sync(); err != nil {
		return err
	}
	return nil
}

// AppendEvent is a compatibility spelling for Append.
func (j *FileJournal) AppendEvent(event Event) error {
	return j.Append(event)
}

// ReadAll reads and validates every JSONL record. A malformed line is an
// integrity error; silently skipping it would make recovery state ambiguous.
func (j *FileJournal) ReadAll() ([]Event, error) {
	if j == nil {
		return nil, errors.New("nil journal")
	}
	if j.path == "" {
		return nil, errors.New("journal path is empty")
	}
	j.mu.Lock()
	defer j.mu.Unlock()
	if err := rejectJournalSymlink(j.path); err != nil {
		return nil, err
	}
	file, err := os.Open(j.path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return []Event{}, nil
		}
		return nil, err
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	// A normal event is small, but permit a bounded diagnostic reason without
	// allowing an unbounded scanner allocation.
	scanner.Buffer(make([]byte, 4096), 1024*1024)
	events := make([]Event, 0)
	line := 0
	for scanner.Scan() {
		line++
		data := bytes.TrimSpace(scanner.Bytes())
		if len(data) == 0 {
			continue
		}
		var event Event
		if err := json.Unmarshal(data, &event); err != nil {
			return nil, fmt.Errorf("%w: line %d: %v", ErrCorruptJournal, line, err)
		}
		if err := ValidateEvent(event); err != nil {
			return nil, fmt.Errorf("%w: line %d: %v", ErrCorruptJournal, line, err)
		}
		events = append(events, event)
	}
	if err := scanner.Err(); err != nil {
		return nil, fmt.Errorf("%w: line %d: %v", ErrCorruptJournal, line+1, err)
	}
	return events, nil
}

// Events is an alias for ReadAll.
func (j *FileJournal) Events() ([]Event, error) {
	return j.ReadAll()
}

// ValidateEvent enforces the scalar event contract and rejects obvious secret
// field patterns in operator-provided text. Raw credential values cannot be
// represented as structured fields in Event.
func ValidateEvent(event Event) error {
	if strings.TrimSpace(string(event.Type)) == "" {
		return ErrInvalidEvent
	}
	if len(event.Type) > 256 {
		return fmt.Errorf("%w: event type is too long", ErrInvalidEvent)
	}
	if event.At.IsZero() {
		return fmt.Errorf("%w: timestamp is required", ErrInvalidEvent)
	}
	if isTransitionEvent(event.Type) {
		if event.TransitionID == "" {
			return fmt.Errorf("%w: transition_id is required for %s", ErrInvalidEvent, event.Type)
		}
		if event.RuntimeID == "" {
			return fmt.Errorf("%w: runtime_id is required for %s", ErrInvalidEvent, event.Type)
		}
		if (event.Type == TransitionCommitted || event.Type == TransitionReconciled) && event.AuthGeneration == 0 {
			return fmt.Errorf("%w: auth_generation is required for %s", ErrInvalidEvent, event.Type)
		}
	}
	if isRecoveryEvent(event.Type) {
		if event.RecoveryID == "" {
			return fmt.Errorf("%w: recovery_id is required for %s", ErrInvalidEvent, event.Type)
		}
		if event.RuntimeID == "" {
			return fmt.Errorf("%w: runtime_id is required for %s", ErrInvalidEvent, event.Type)
		}
	}
	for name, value := range map[string]string{
		"transition_id":     event.TransitionID,
		"recovery_id":       event.RecoveryID,
		"thread_id":         event.ThreadID,
		"turn_id":           event.TurnID,
		"runtime_id":        event.RuntimeID,
		"account_id":        event.AccountID,
		"from_account_id":   event.FromAccountID,
		"target_account_id": event.TargetAccountID,
		"outcome":           event.Outcome,
		"reason":            event.Reason,
	} {
		if err := validateText(name, value); err != nil {
			return err
		}
	}
	return nil
}

func isTransitionEvent(eventType EventType) bool {
	switch eventType {
	case TransitionCreated, TransitionPrepared, AuthDeployed, CommitSent,
		IdentityConfirmed, TransitionCommitted, TransitionUncertain,
		TransitionReconciled:
		return true
	default:
		return false
	}
}

func isRecoveryEvent(eventType EventType) bool {
	switch eventType {
	case RecoveryParked, RecoveryWaiting, RecoveryBound,
		RecoveryReleaseRequested, RecoveryReleased, RecoveryStarted,
		RecoveryCompleted, RecoveryUncertain:
		return true
	default:
		return false
	}
}

func validateText(field, value string) error {
	if strings.ContainsRune(value, 0) || strings.ContainsAny(value, "\r\n") {
		return fmt.Errorf("%w: %s contains control characters", ErrInvalidEvent, field)
	}
	if len(value) > 4096 {
		return fmt.Errorf("%w: %s is too long", ErrInvalidEvent, field)
	}
	if field == "reason" && looksSensitive(value) {
		return fmt.Errorf("%w: %s", ErrSensitiveEvent, field)
	}
	return nil
}

func looksSensitive(value string) bool {
	lower := strings.ToLower(value)
	for _, marker := range []string{
		"access_token", "id_token", "refresh_token", "authorization:",
		"bearer ", "api_key", "openai_api_key", "password=", "secret=",
		"auth.json", "\"tokens\"",
	} {
		if strings.Contains(lower, marker) {
			return true
		}
	}
	return false
}

func isPermissionMetadataError(error) bool {
	return false
}

func rejectJournalSymlink(path string) error {
	info, err := os.Lstat(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil
		}
		return err
	}
	if info.Mode()&os.ModeSymlink != 0 {
		return fmt.Errorf("%w: journal path is a symbolic link", ErrInvalidEvent)
	}
	if info.IsDir() {
		return fmt.Errorf("%w: journal path is a directory", ErrInvalidEvent)
	}
	return nil
}

File: runtime/codex-rs/codexmarathon-runtime/src/journal.rs
//! Append-only, metadata-only transition journal.
//!
//! Journal records intentionally have no field capable of carrying an auth
//! snapshot, token, authorization header, or recovery prompt. Validation also
//! rejects common secret markers in the free-form reason field.

use crate::errors::{DomainError, DomainResult};
use crate::persistence::{ensure_private_dir, reject_unsafe_file, set_private_permissions_file};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::fs::{File, OpenOptions};
use std::io::{BufRead, BufReader, Write};
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex};

/// Stable journal event type names used by transition and recovery code.
pub mod event_types {
    pub const TRANSITION_CREATED: &str = "TransitionCreated";
    pub const TRANSITION_PREPARED: &str = "TransitionPrepared";
    pub const AUTH_DEPLOYED: &str = "AuthDeployed";
    pub const COMMIT_SENT: &str = "CommitSent";
    pub const IDENTITY_CONFIRMED: &str = "IdentityConfirmed";
    pub const TRANSITION_COMMITTED: &str = "TransitionCommitted";
    pub const TRANSITION_UNCERTAIN: &str = "TransitionUncertain";
    pub const TRANSITION_RECONCILED: &str = "TransitionReconciled";
    pub const RUNTIME_CONNECTED: &str = "RuntimeConnected";
    pub const RUNTIME_DISCONNECTED: &str = "RuntimeDisconnected";
    pub const RECOVERY_PARKED: &str = "RecoveryParked";
    pub const RECOVERY_WAITING: &str = "RecoveryWaiting";
    pub const RECOVERY_BOUND: &str = "RecoveryBound";
    pub const RECOVERY_RELEASE_REQUESTED: &str = "RecoveryReleaseRequested";
    pub const RECOVERY_RELEASED: &str = "RecoveryReleased";
    pub const RECOVERY_STARTED: &str = "RecoveryStarted";
    pub const RECOVERY_COMPLETED: &str = "RecoveryCompleted";
    pub const RECOVERY_UNCERTAIN: &str = "RecoveryUncertain";
}

/// A scalar-only journal record.
#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct JournalEvent {
    /// Event timestamp in UTC.
    pub at: DateTime<Utc>,
    /// Stable event type name.
    #[serde(rename = "type")]
    pub event_type: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub transition_id: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub recovery_id: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub thread_id: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub turn_id: String,
    /// Runtime auth generation observed for this event.
    pub auth_generation: u64,
    #[serde(default, skip_serializing_if = "is_zero")]
    pub expected_generation: u64,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub runtime_id: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub account_id: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub from_account_id: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub target_account_id: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub outcome: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub reason: String,
}

fn is_zero(value: &u64) -> bool {
    *value == 0
}

impl JournalEvent {
    /// Construct an event with the current timestamp and empty metadata.
    pub fn new(event_type: impl Into<String>) -> Self {
        Self {
            at: Utc::now(),
            event_type: event_type.into(),
            transition_id: String::new(),
            recovery_id: String::new(),
            thread_id: String::new(),
            turn_id: String::new(),
            auth_generation: 0,
            expected_generation: 0,
            runtime_id: String::new(),
            account_id: String::new(),
            from_account_id: String::new(),
            target_account_id: String::new(),
            outcome: String::new(),
            reason: String::new(),
        }
    }

    /// Validate the metadata-only event contract.
    pub fn validate(&self) -> DomainResult<()> {
        if self.event_type.trim().is_empty() || self.event_type.len() > 256 {
            return Err(DomainError::InvalidJournalEvent);
        }
        if self.at.timestamp() == 0 && self.at.timestamp_subsec_nanos() == 0 {
            return Err(DomainError::InvalidJournalEvent);
        }
        if is_transition_event(&self.event_type) {
            if self.transition_id.is_empty() || self.runtime_id.is_empty() {
                return Err(DomainError::InvalidJournalEvent);
            }
            if matches!(
                self.event_type.as_str(),
                event_types::TRANSITION_COMMITTED | event_types::TRANSITION_RECONCILED
            ) && self.auth_generation == 0
            {
                return Err(DomainError::InvalidJournalEvent);
            }
        }
        if is_recovery_event(&self.event_type)
            && (self.recovery_id.is_empty() || self.runtime_id.is_empty())
        {
            return Err(DomainError::InvalidJournalEvent);
        }
        for value in [
            &self.transition_id,
            &self.recovery_id,
            &self.thread_id,
            &self.turn_id,
            &self.runtime_id,
            &self.account_id,
            &self.from_account_id,
            &self.target_account_id,
            &self.outcome,
            &self.reason,
        ] {
            if value.len() > 4096
                || value
                    .chars()
                    .any(|character| character == '\0' || character.is_control())
            {
                return Err(DomainError::InvalidJournalEvent);
            }
        }
        if looks_sensitive(&self.reason) {
            return Err(DomainError::SensitiveJournalEvent);
        }
        Ok(())
    }
}

fn is_transition_event(event_type: &str) -> bool {
    matches!(
        event_type,
        event_types::TRANSITION_CREATED
            | event_types::TRANSITION_PREPARED
            | event_types::AUTH_DEPLOYED
            | event_types::COMMIT_SENT
            | event_types::IDENTITY_CONFIRMED
            | event_types::TRANSITION_COMMITTED
            | event_types::TRANSITION_UNCERTAIN
            | event_types::TRANSITION_RECONCILED
    )
}

fn is_recovery_event(event_type: &str) -> bool {
    matches!(
        event_type,
        event_types::RECOVERY_PARKED
            | event_types::RECOVERY_WAITING
            | event_types::RECOVERY_BOUND
            | event_types::RECOVERY_RELEASE_REQUESTED
            | event_types::RECOVERY_RELEASED
            | event_types::RECOVERY_STARTED
            | event_types::RECOVERY_COMPLETED
            | event_types::RECOVERY_UNCERTAIN
    )
}

fn looks_sensitive(value: &str) -> bool {
    let value = value.to_ascii_lowercase();
    [
        "access_token",
        "id_token",
        "refresh_token",
        "authorization:",
        "bearer ",
        "api_key",
        "openai_api_key",
        "password=",
        "secret=",
        "auth.json",
        "\"tokens\"",
    ]
    .iter()
    .any(|marker| value.contains(marker))
}

/// Storage-neutral journal interface.
pub trait Journal: Send + Sync {
    /// Validate and durably append an event.
    fn append(&self, event: JournalEvent) -> DomainResult<()>;
    /// Read and validate all events in order.
    fn read_all(&self) -> DomainResult<Vec<JournalEvent>>;
}

/// Process-safe JSONL journal with owner-only permissions.
pub struct FileJournal {
    path: PathBuf,
    gate: Mutex<()>,
    clock: Arc<dyn Fn() -> DateTime<Utc> + Send + Sync>,
}

impl FileJournal {
    /// Construct a journal without creating its file.
    pub fn new(path: impl Into<PathBuf>) -> Self {
        Self::with_clock(path, Arc::new(Utc::now))
    }

    /// Construct a journal with an injectable clock.
    pub fn with_clock(
        path: impl Into<PathBuf>,
        clock: Arc<dyn Fn() -> DateTime<Utc> + Send + Sync>,
    ) -> Self {
        Self {
            path: path.into(),
            gate: Mutex::new(()),
            clock,
        }
    }

    /// Return the journal path without reading records.
    pub fn path(&self) -> &Path {
        &self.path
    }
}

impl Journal for FileJournal {
    fn append(&self, mut event: JournalEvent) -> DomainResult<()> {
        let _guard = self.gate.lock().map_err(|_| DomainError::CorruptJournal)?;
        if event.at.timestamp() == 0 && event.at.timestamp_subsec_nanos() == 0 {
            event.at = (self.clock)();
        }
        event.at = event.at.with_timezone(&Utc);
        event.validate()?;
        reject_unsafe_file(&self.path)?;
        let parent = self.path.parent().ok_or(DomainError::UnsafePath)?;
        ensure_private_dir(parent)?;
        let mut file = OpenOptions::new()
            .create(true)
            .append(true)
            .open(&self.path)?;
        set_private_permissions_file(&file)?;
        serde_json::to_writer(&mut file, &event)?;
        file.write_all(b"\n")?;
        file.sync_all()?;
        Ok(())
    }

    fn read_all(&self) -> DomainResult<Vec<JournalEvent>> {
        let _guard = self.gate.lock().map_err(|_| DomainError::CorruptJournal)?;
        reject_unsafe_file(&self.path)?;
        let file = match File::open(&self.path) {
            Ok(file) => file,
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => return Ok(Vec::new()),
            Err(error) => return Err(error.into()),
        };
        let mut records = Vec::new();
        let reader = BufReader::new(file);
        for line in reader.lines() {
            let line = line?;
            if line.trim().is_empty() {
                continue;
            }
            if line.len() > 1024 * 1024 {
                return Err(DomainError::CorruptJournal);
            }
            let event: JournalEvent =
                serde_json::from_str(&line).map_err(|_| DomainError::CorruptJournal)?;
            event.validate().map_err(|_| DomainError::CorruptJournal)?;
            records.push(event);
        }
        Ok(records)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn journal_round_trip_and_secret_guard() {
        let directory = tempdir().expect("tempdir");
        let journal = FileJournal::new(directory.path().join("journal.jsonl"));
        let mut event = JournalEvent::new(event_types::TRANSITION_CREATED);
        event.transition_id = "transition-1".to_string();
        event.runtime_id = "runtime-1".to_string();
        event.reason = "threshold reached".to_string();
        journal.append(event.clone()).expect("append");
        assert_eq!(journal.read_all().expect("read"), vec![event]);

        let mut sensitive = JournalEvent::new("diagnostic");
        sensitive.reason = "Authorization: Bearer secret".to_string();
        assert!(matches!(
            journal.append(sensitive),
            Err(DomainError::SensitiveJournalEvent)
        ));
    }

    #[test]
    fn committed_transition_requires_generation() {
        let mut event = JournalEvent::new(event_types::TRANSITION_COMMITTED);
        event.transition_id = "transition-1".to_string();
        event.runtime_id = "runtime-1".to_string();
        assert!(matches!(
            event.validate(),
            Err(DomainError::InvalidJournalEvent)
        ));
        event.auth_generation = 2;
        assert!(event.validate().is_ok());
    }
}
